runtime/grid_reduction.cu

```diff
@@ -667,4 +667,62 @@ __device__ void serialReductionStep(
   }
 }
 
+// Same as above, but cast to lower precision for work buffer IO
+template <int64_t vec_size, typename Twork, typename Func>
+__device__ void serialReductionStep(
+    float* out,
+    float* in,
+    float init,
+    volatile Twork* work,
```

Comment:
This template's redundant since `Twork = float` might match this if we're not careful. Instead it might be best to just check `is_same` for the types of `out` and `work` and dispatch from there to separately-named helper functions.